state/application_test.go

```diff
@@ -922,6 +922,209 @@ func (s *ApplicationSuite) TestUpdateConfigSettings(c *gc.C) {
 	}
 }
 
+func (s *ApplicationSuite) TestUpdateApplicationSeries(c *gc.C) {
+	ch := state.AddTestingCharmMultiSeries(c, s.State, "multi-series")
+	app := state.AddTestingServiceForSeries(c, s.State, "precise", "multi-series", ch)
+	err := app.UpdateApplicationSeries("trusty", false)
+	c.Assert(err, jc.ErrorIsNil)
+	app.Refresh()
+	c.Assert(app.Series(), gc.Equals, "trusty")
+}
+
+func (s *ApplicationSuite) setupMultiSeriesUnitWithSubordinate(c *gc.C) (*state.Application, *state.Application) {
+	ch := state.AddTestingCharmMultiSeries(c, s.State, "multi-series")
+	app := state.AddTestingServiceForSeries(c, s.State, "precise", "multi-series", ch)
+	subCh := state.AddTestingCharmMultiSeries(c, s.State, "multi-series-subordinate")
+	subApp := state.AddTestingServiceForSeries(c, s.State, "precise", "multi-series-subordinate", subCh)
+
+	eps, err := s.State.InferEndpoints("multi-series", "multi-series-subordinate")
+	c.Assert(err, jc.ErrorIsNil)
+	rel, err := s.State.AddRelation(eps...)
+	c.Assert(err, jc.ErrorIsNil)
+
+	unit, err := app.AddUnit()
+	c.Assert(err, jc.ErrorIsNil)
+	err = unit.AssignToNewMachine()
+	c.Assert(err, jc.ErrorIsNil)
+
+	ru, err := rel.Unit(unit)
+	c.Assert(err, jc.ErrorIsNil)
+	err = ru.EnterScope(nil)
+	c.Assert(err, jc.ErrorIsNil)
+
+	app.Refresh()
+	subApp.Refresh()
+
+	return app, subApp
+}
+
+func (s *ApplicationSuite) TestUpdateApplicationSeriesWithSubordinate(c *gc.C) {
+	app, subApp := s.setupMultiSeriesUnitWithSubordinate(c)
+	err := app.UpdateApplicationSeries("trusty", false)
+	c.Assert(err, jc.ErrorIsNil)
+	app.Refresh()
+	c.Assert(app.Series(), gc.Equals, "trusty")
+	subApp.Refresh()
+	c.Assert(subApp.Series(), gc.Equals, "trusty")
+}
+
+func (s *ApplicationSuite) TestUpdateApplicationSeriesWithSubordinateFail(c *gc.C) {
+	app, subApp := s.setupMultiSeriesUnitWithSubordinate(c)
+	err := app.UpdateApplicationSeries("xenial", false)
+	c.Assert(err, jc.Satisfies, state.IsIncompatibleSeriesError)
+	app.Refresh()
+	c.Assert(app.Series(), gc.Equals, "precise")
+	subApp.Refresh()
+	c.Assert(subApp.Series(), gc.Equals, "precise")
+}
+
+func (s *ApplicationSuite) TestUpdateApplicationSeriesWithSubordinateForce(c *gc.C) {
+	app, subApp := s.setupMultiSeriesUnitWithSubordinate(c)
+	err := app.UpdateApplicationSeries("xenial", true)
+	c.Assert(err, jc.ErrorIsNil)
+	app.Refresh()
+	c.Assert(app.Series(), gc.Equals, "xenial")
+	subApp.Refresh()
+	c.Assert(subApp.Series(), gc.Equals, "xenial")
+}
+
+func (s *ApplicationSuite) TestUpdateApplicationSeriesURLChange(c *gc.C) {
+	ch := state.AddTestingCharmMultiSeries(c, s.State, "multi-series")
+	app := state.AddTestingServiceForSeries(c, s.State, "precise", "multi-series", ch)
+	curl, _ := app.CharmURL()
+
+	defer state.SetTestHooks(c, s.State,
+		jujutxn.TestHook{
+			Before: func() {
+				ops := []txn.Op{{
```

Comment:
We typically just manipulate state directly via state.State api calls. Using the txn runner works but should be a last resort. eg could just use app.SetCharm(blah)